pkg/cloudprovider/providers/vsphere/vsphere_util.go

```diff
@@ -63,3 +81,226 @@ func GetgovmomiClient(cfg *VSphereConfig) (*govmomi.Client, error) {
+	datastores, err := vs.getSharedDatastoresInK8SCluster(ctx)
+	if err != nil {
+		return nil, err
+	}
+	var hubs []pbmtypes.PbmPlacementHub
+	for _, ds := range datastores {
+		hubs = append(hubs, pbmtypes.PbmPlacementHub{
+			HubType: ds.Type,
+			HubId:   ds.Value,
+		})
+	}
+	req := []pbmtypes.BasePbmPlacementRequirement{
+		&pbmtypes.PbmPlacementCapabilityProfileRequirement{
+			ProfileId: pbmtypes.PbmProfileId{
+				UniqueId: storagePolicyID,
+			},
+		},
+	}
+	res, err := pbmClient.CheckRequirements(ctx, hubs, nil, req)
+	if err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// Verify if the user specified datastore is in the list of non-compatible datastores.
+// If yes, return the non compatible datastore reference.
+func (vs *VSphere) IsUserSpecifiedDatastoreNonCompatible(ctx context.Context, compatibilityResult pbm.PlacementCompatibilityResult, dsName string) (bool, *types.ManagedObjectReference) {
+	dsMoList := vs.GetNonCompatibleDatastoresMo(ctx, compatibilityResult)
+	for _, ds := range dsMoList {
+		if ds.Info.GetDatastoreInfo().Name == dsName {
+			dsMoRef := ds.Reference()
+			return true, &dsMoRef
+		}
+	}
+	return false, nil
+}
+
+func GetNonCompatibleDatastoreFaultMsg(compatibilityResult pbm.PlacementCompatibilityResult, dsMoref types.ManagedObjectReference) string {
+	var faultMsg string
+	for _, res := range compatibilityResult {
+		if res.Hub.HubId == dsMoref.Value {
+			for _, err := range res.Error {
+				faultMsg = faultMsg + err.LocalizedMessage
+			}
+		}
+	}
+	return faultMsg
+}
+
+// Get the best fit compatible datastore by free space.
+func GetMostFreeDatastore(dsMo []mo.Datastore) mo.Datastore {
+	var curMax int64
+	curMax = -1
+	var index int
+	for i, ds := range dsMo {
+		dsFreeSpace := ds.Info.GetDatastoreInfo().FreeSpace
+		if dsFreeSpace > curMax {
+			curMax = dsFreeSpace
+			index = i
+		}
+	}
+	return dsMo[index]
+}
+
+func (vs *VSphere) GetCompatibleDatastoresMo(ctx context.Context, compatibilityResult pbm.PlacementCompatibilityResult) ([]mo.Datastore, error) {
+	compatibleHubs := compatibilityResult.CompatibleDatastores()
+	// Return an error if there are no compatible datastores.
+	if len(compatibleHubs) < 1 {
+		return nil, fmt.Errorf("There are no compatible datastores that satisfy the storage policy requirements")
+	}
+	dsMoList, err := vs.getDatastoreMo(ctx, compatibleHubs)
+	if err != nil {
+		return nil, err
+	}
+	return dsMoList, nil
+}
+
+func (vs *VSphere) GetNonCompatibleDatastoresMo(ctx context.Context, compatibilityResult pbm.PlacementCompatibilityResult) []mo.Datastore {
+	nonCompatibleHubs := compatibilityResult.NonCompatibleDatastores()
+	// Return an error if there are no compatible datastores.
+	if len(nonCompatibleHubs) < 1 {
+		return nil
+	}
+	dsMoList, err := vs.getDatastoreMo(ctx, nonCompatibleHubs)
+	if err != nil {
+		return nil
```

Comment:
No, GetNonCompatibleDatastoresMo returns datastore list if there are any, otherwise nil.  Check the return type.